File: solutions/dart/protein-translation/1/lib/protein_translation.dart
// Dedicated to Shree DR.MDD 🙏

class ProteinTranslation {
  static const sentinel = 'stop';
  static const error = 'invalid';

  List<String> translate(String rnaStrand) {
    final peptideChain = RegExp('.{1,3}')
        .allMatches(rnaStrand)
        .map((codonMatch) => _decode(codonMatch[0]!))
        .takeWhile((amino) => amino != sentinel);

    if (peptideChain.contains(error)) {
      throw ArgumentError('invalid codon');
    }

    return peptideChain.toList();
  }

  String _decode(String triplet) {
    switch (triplet) {
      case 'AUG':
        return 'Methionine';
      case 'UUU':
      case 'UUC':
        return 'Phenylalanine';
      case 'UUA':
      case 'UUG':
        return 'Leucine';
      case 'UCU':
      case 'UCC':
      case 'UCA':
      case 'UCG':
        return 'Serine';
      case 'UAU':
      case 'UAC':
        return 'Tyrosine';
      case 'UGU':
      case 'UGC':
        return 'Cysteine';
      case 'UGG':
        return 'Tryptophan';
      case 'UAA':
      case 'UAG':
      case 'UGA':
        return sentinel;
      default:
        return error;
    }
  }
}

File: solutions/php/protein-translation/1/ProteinTranslation.php
<?php
// For my Shree DR.MDD

declare(strict_types=1);

class ProteinTranslation
{
    private array $codonTable = [
        'AUG' => 'Methionine',
        'UUU' => 'Phenylalanine',
        'UUC' => 'Phenylalanine',
        'UUA' => 'Leucine',
        'UUG' => 'Leucine',
        'UCU' => 'Serine',
        'UCC' => 'Serine',
        'UCA' => 'Serine',
        'UCG' => 'Serine',
        'UAU' => 'Tyrosine',
        'UAC' => 'Tyrosine',
        'UGU' => 'Cysteine',
        'UGC' => 'Cysteine',
        'UGG' => 'Tryptophan',
        'UAA' => 'STOP',
        'UAG' => 'STOP',
        'UGA' => 'STOP',
    ];

    public function getProteins(string $strand): array
    {
        $triplets = str_split($strand, 3);
        $proteins = [];
        foreach ($triplets as $triplet) {
            $amino = $this->codonTable[$triplet] ?? NULL;
            if ($amino === NULL) throw new InvalidArgumentException('Invalid codon');
            if ($amino === "STOP") return $proteins;
            $proteins[] = $amino;
        }
        return $proteins;
    }
}
